tachyon/c/zk/plonk/halo2/buffer_reader.h

```diff
@@ -405,6 +407,43 @@ class BufferReader<tachyon::zk::plonk::PermutationProvingKey<Poly, Evals>> {
   }
 };
 
+template <typename T>
+class BufferReader<
+    T, std::enable_if_t<
+           std::is_same_v<T, c::zk::plonk::halo2::bn254::GWCPCS> ||
+           std::is_same_v<T, c::zk::plonk::halo2::bn254::SHPlonkPCS>>> {
+ public:
+  static T Read(const tachyon::base::ReadOnlyBuffer& buffer) {
+    using G1Point = typename T::G1Point;
+    using G2Point = typename T::G2Point;
+
+    uint32_t k;
+    CHECK(buffer.Read(&k));
+    size_t n = size_t{1} << k;
+    std::vector<G1Point> g1_powers_of_tau =
+        tachyon::base::CreateVector(n, [&buffer]() {
+          G1Point point;
+          c::zk::plonk::ReadBuffer(buffer, point);
+          return point;
+        });
+    std::vector<G1Point> g1_powers_of_tau_lagrange =
+        tachyon::base::CreateVector(n, [&buffer]() {
+          G1Point point;
+          c::zk::plonk::ReadBuffer(buffer, point);
+          return point;
+        });
+
+    // NOTE(dongchangYoo): read |g2| but do not use it.
+    G2Point g2;
+    c::zk::plonk::ReadBuffer(buffer, g2);
+
+    G2Point s_g2;
+    c::zk::plonk::ReadBuffer(buffer, s_g2);
```

Comment:
I guess you don't need `c::zk::plonk` cuz its in the same namespace.
```suggestion
	  ReadBuffer(buffer, point);
          return point;
        });
    std::vector<G1Point> g1_powers_of_tau_lagrange =
        tachyon::base::CreateVector(n, [&buffer]() {
          G1Point point;
          ReadBuffer(buffer, point);
          return point;
        });

    // NOTE(dongchangYoo): read |g2| but do not use it.
    G2Point g2;
    ReadBuffer(buffer, g2);

    G2Point s_g2;
    ReadBuffer(buffer, s_g2);
```